### DISCUSSION TITLE
[Stream] SAC Active and Waiting Comumers both must be able to listen event.

### DISCUSSION THREAD

**Discussion by laststem** [rabbitmq/rabbitmq-server#11184] [Ideas] [answered]
I'm not sure this repository is suitable to discuss this topic.

in the https://github.com/rabbitmq/rabbitmq-stream-java-client, there is `ConsumerUpdateListener`.

I expected to any update event(regardless of active, waiting) will invoke `ConsumerUpdateListener`.
because there is `ConsumerUpdateListener.Context.isActive()` that can inform this consumer is active(isActive()=true) or waiting(isActive()=false).

but, it invoked `active(isActive()=true)` consumer process only.

```kotlin
        val consumer = env.consumerBuilder()
            .name("test")
            .stream("playground")
            .singleActiveConsumer()
            .messageHandler { context, message ->
                println("offset: ${context.offset()}, message: ${String(message.bodyAsBinary)}")
            }
            .consumerUpdateListener { ctx ->
                println("UpdateListener: $ctx")
                OffsetSpecification.last()
            }
            .build()
```

The above code prints `UpdateListener: $ctx` only once even if there are multiple processes.

Is it Stream SAC specification that only call active consumer?

**Comment by michaelklishin** [comment]
It's a philosophical question, whether events should be broadcast to all consumers or just the active one if the SAC mode is used. I'd say only the active one. If you want N active consumers then don't use the SAC feature.

There may be situation where some metadata updates exposed to the user (your app) might be worth distributing to all consumers but you have to come up with a specific example. SAC tries to create an illusion that there's only one consumer online.

**Reply by laststem** [reply]
sorry for not enough explanation.

I didn't mean to broadcast events to all consumers. but events should be delivered when a consumer is registered.

When an event is delivered when each consumer registers with a stream, the consumer application can determine whether it is a 'leader (active) application' or a 'follower (standby) application'.

Currently consumer applications do not know whether they are 'follower applications'. This is because the broker does not send update events to the waiting consumer except the active application.

The specific scenario is as follows.

1. Consumers do not exist.
2. A consumer was registered.
3. Broker send update event (A is active)
4. B consumer was registered.
5. A got event(active=true), execute Leader logic.
6. Broker send update event (B is waiting)
7. B got event(active=false), execute Follower logic.

If the broker does not send an event to B, B cannot know whether it is not receiving the event because the connection is still in progress or whether it is a follower.

**Reply by michaelklishin** [reply]
I see your point that applications would need a hint of sorts. We will discuss what options do we have for connections/consumers within a single consumer group.

**Comment by michaelklishin** [comment]
From a team member who was one of the key contributors to the SAC feature:

```
The consumer update listener is triggered only when there's a change for the consumer.
Events could in theory be sent to the entire SAC member group but then consumers
would receive events that do not apply to them.
```

**Comment by acogoluegnes** [comment, marked as answer]
In a single active consumer group, a consumer is considered inactive by default, until it receives a notification that it is active. Then an active consumer can receive a notification that it becomes inactive.

A consumer gets to decide where to start consuming from when it becomes active: it responds to the notification with an offset specification and the broker will use it to initialize the message flow at the appropriate location in the stream. It can get the offset notification from any data source (RabbitMQ with server-side offset tracking or a custom database).

An active consumer has a chance to store the last offset it processed when it gets a notification that it becomes inactive. This way the new active consumer will be able to resume where the former active consumer left off.

The frequency of the consumer update notification is meant to make the system work, not to broadcast events to consumers. We do not plan on changing this unless there is a good reason to do it.

**Reply by laststem** [reply]
SAC active consumer can't receive a notification (active -> inactive)

Reproducible scenario)
1. Broker 3 exist.
2. Stream A exist. (Leader=Broker0, Replica=Broker1,Broker2)
3. Consumer A subscribes Stream A
4. Consumer A connected Broker 1 and got SAC event(active=true)
5. Consumer B connected Broker 1 (without SAC event)
6. Broker 1 restart
7. Consumer B got SAC event(active=true)

Please provide can receive notification (active -> inactive) scenario.
@acogoluegnes @michaelklishin 
Thanks.

**Reply by michaelklishin** [reply]
@laststem I think this answers your questions

> In a single active consumer group, a consumer is considered inactive by default, until it receives a notification that it is active

With a node restart, all consumers previously connected will be disconnected, then reconnect and only the active one will receive a notifications,
and others will be considered passive.

**Reply by laststem** [reply]
Okay.

for simple use cases, it might be better to default to a waiting consumer.
but for complex use cases, it might be better to receive an inactive event.
So, consumer application can recognize I am active or inactive.
If it was initially active and then became inactive, it cannot be recognized because there is no callback provided by the library.

I think it would be nice to also deliver events for inactive.
And I don't think there are any downsides to doing so.

**Reply by acogoluegnes** [reply]
I have to test, but it is possible an active consumer does not receive an "inactive" notification if it gets disconnected from the broker and tries to recover. The broker would not send anything, but the library could enforce the callback. There is no way to tell the consumer will be active again, so it's safer to mimick an "inactive" notification.

> Please provide can receive notification (active -> inactive) scenario.

I'm not sure it is a question and if I understand it correctly, but an active consumer can get deactivated in favor of another consumer when a new consumer joins the group.

**Reply by laststem** [reply]
Thanks for reply.

> but it is possible an active consumer does not receive an "inactive" notification if it gets disconnected from the broker

I agree, so consumer couldn't receive inactive notification by broker.

> but the library could enforce the callback.

Is the current library implemented to pass inactive instead of broker?
I've tested it several times but I've never received an inactive notification.

> but an active consumer can get deactivated in favor of another consumer when a new consumer joins the group.

I think this won't happen if there are no problems with existing active consumers, right?

**Reply by acogoluegnes** [reply]
> 
> > but the library could enforce the callback.
> 
> Is the current library implemented to pass inactive instead of broker? I've tested it several times but I've never received an inactive notification.
> 

The library does not implement this. I'll have a closer look and see if it's appropriate to change this behavior, by making the library enforce the callback, even though the broker cannot send anything because the connection has been dropped.

> > but an active consumer can get deactivated in favor of another consumer when a new consumer joins the group.
> 
> I think this won't happen if there are no problems with existing active consumers, right?

It should not happen for a SAC on a single stream, it can happen when using SAC with super streams. Sorry for the confusion.

**Reply by acogoluegnes** [reply]
@laststem I pushed a change to notify a SAC that goes from active to inactive for cases the broker does not do so. Can try out the [latest snapshot](https://rabbitmq.github.io/rabbitmq-stream-java-client/snapshot/htmlsingle/#dependencies) to check it behaves as expected? Thanks.

**Reply by laststem** [reply]
0.18.0-SNAPSHOT seems to works well. Thanks @acogoluegnes

**Reply by acogoluegnes** [reply]
OK, thanks for confirming.

**Reply by laststem** [reply]
When will this be released? @acogoluegnes thanks

**Reply by acogoluegnes** [reply]
We will release RabbitMQ Stream Java client 0.18.0 this week.
